kbase-extension/static/kbase/js/common/simpleLogViewer.js

```diff
@@ -0,0 +1,818 @@
+                        case 'queued':
+                            this.ui.setContent(LOG_PANEL, p([this.messages.JOB_QUEUED]));
+                            this._renderButtonState('default');
+                            break;
+                        case 'running':
+                            // clear the content of the job panel to make way for logs
+                            this.ui.setContent(LOG_PANEL, '');
+                            this.startLogAutoFetch();
+                            this._renderButtonState('running_scrolling');
+                            break;
+                        case 'terminal':
+                            this.stopLogAutoFetch();
+                            this.requestJobLog();
+                            this._renderButtonState('terminal');
+                            break;
+                        // this should never happen
+                        default:
+                            console.error('Unknown job status', status, message);
+                            throw new Error(`Unknown job status ${status}`);
+                    }
+                }
+            }
+
+            handleJobLogs(message) {
+                if (!this.state.awaitingLog) {
+                    return;
+                }
+
+                this.ui.hideElement('spinner');
+                /* message has structure:
+                 * {
+                 *      first: int (first line of the log batch), 0-indexed
+                 *      job_id: string,
+                 *      latest: bool,
+                 *      max_lines: int (total logs available - if job is done, so is this),
+                 *      lines: [{
+                 *          is_error: 0 or 1,
+                 *          line: string,
+                 *          linepos: int, position in log. helpful!
+                 *          ts: timestamp
+                 *      }]
+                 * }
+                 */
+                this.state.awaitingLog = false;
+
+                if (message.error) {
+                    return this.handleLogError(message);
+                }
+                if (this.state.looping) {
+                    this.scheduleLogRequest();
+                }
+
+                if (message.lines.length !== 0) {
+                    const savedLogs = this.model.getRawObject();
+                    // ensure that message.first is equal to or less than savedLogs.lastLine
+                    if (!savedLogs.lastLine || message.first <= savedLogs.lastLine) {
+                        // no need to update if we already have all the lines
+                        if (message.first + message.lines.length <= savedLogs.lastLine) {
+                            return;
+                        }
+                        savedLogs.lines.splice(
+                            message.first,
+                            message.lines.length,
+                            ...message.lines
+                        );
```

Comment:
new -- rather than replacing the whole set of log lines every time, we now just add in any new lines from the log updates.